File: 프로그래머스/4/42891. 무지의 먹방 라이브/무지의 먹방 라이브.py
# 파이썬 1초에 2000만
# 이분탐색 O(logN)
# 200,000 * log(100,000,000) = 5,200,000
def solution(food_times, k):
    left, right = 1, int(1e8)
    N = len(food_times)
    rotate, t = 0, 0
    times = sorted(food_times)
    while left <= right:
        mid = (left + right) // 2
        temp = N * mid
        for time in times:
            if time >= mid:
                break
            temp -= mid - time
        
        if temp <= k:
            rotate = mid
            t = temp
            left = mid + 1
        else:
            right = mid - 1
    
    k -= t
    for i in range(len(food_times)):
        food_times[i] -= rotate
        if food_times[i] > 0:
            k -= 1
        if k < 0:
            return i + 1

    return -1

File: 프로그래머스/4/42891. 무지의 먹방 라이브/무지의 먹방 라이브.java
import java.util.*;

class Solution {
    public int solution(int[] food_times, long k) {
        long left = 1;
        long right = Arrays.stream(food_times).max().getAsInt();
        long t = 0;
        long eat = 0;
        while (left <= right) {
            long mid = (left + right) / 2;
            long temp = 0;
            for (int time: food_times) {
                temp += Math.min(mid, time);
            }
            if (temp <= k) {
                t = mid;
                eat = temp;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        k -= eat;
        for (int i=0; i<food_times.length; i++) {
            if (food_times[i] > t) {
                if (k == 0) {
                    return i+1;
                }
                k--;
            }
        }
        return -1;
    }
}